tests/system/runtimes/test_nuclio.py

```diff
@@ -582,3 +582,76 @@ def test_hyper_run(self):
         assert run_result.state() == "completed", "wrong state"
         # accuracy = max(p1) * 2, stop where accuracy > 9
         assert run_result.output("accuracy") == 10, "unexpected results"
+
+
+@tests.system.base.TestMLRunSystem.skip_test_if_env_not_configured
+class TestNuclioAPIGateways(tests.system.base.TestMLRunSystem):
+    project_name = "nuclio-mlrun-gateways"
+    gw_name = "test-gateway"
+
+    def custom_setup(self):
+        self.f1 = self._deploy_function(suffix="1")
+        self.f2 = self._deploy_function(suffix="2")
+
+    def _get_basic_gateway(self):
+        return mlrun.runtimes.nuclio.api_gateway.APIGateway(
+            project=self.project_name, functions=self.f1, name=self.gw_name
+        )
+
+    def _cleanup_gateway(self):
+        self.project.delete_api_gateway(self.gw_name)
+
+    def _deploy_function(self, replicas=1, suffix=""):
+        filename = str(self.assets_path / "nuclio_function.py")
+
+        fn = mlrun.code_to_function(
+            filename=filename,
+            name=f"nuclio-mlrun-{suffix}",
+            kind="nuclio",
+            image="python:3.9",
+            handler="handler",
+        )
+        fn.spec.replicas = replicas
+        fn.with_http(workers=1)
+        fn.deploy()
+        return fn
+
+    def _create_api_gateway_and_wait_for_availability(
+        self, api_gateway, auth=None, max_retries=5, timeout=1
+    ):
+        api_gateway = self.project.store_api_gateway(api_gateway)
+        retry = 0
+        while retry < max_retries:
+            retry += 1
+            result = api_gateway.invoke(auth=auth)
+            if result.status_code == 404:
+                time.sleep(timeout)
+                continue
+            else:
+                return api_gateway
```

Comment:
why wouldnt it be ready once `store` is performed?

Reply:
@liranbg because from nuclio side dashboard returns 201 after CRD creation, but controller should catch this change and create ingress (usually it takes ~1s)

Reply:
I think we should make MLRun API aware of it and ensure it returns the created/updated api gateway only once it is usable (without probing it) so I think we can
1. mlrun api creates/updates api gw on nuclio
2. mlrun api waits until ingress record status block is populated (for nuclio < 1.13)
3. mlrun api returns the api gw



2/3 can be done using by either
1. background tasks
2. return the created entity and expose functionality to "ensure_ingress_readiness()" which will use mlrun api to ensure in the background (short polling)

Reply:
@liranbg reworked